demo/terraform/main.tf

```diff
@@ -0,0 +1,224 @@
+  backend = vault_mount.pki.path
+
+  issuing_certificates    = ["${var.vault_address}/v1/${vault_mount.pki.path}/ca"]
+  crl_distribution_points = ["${var.vault_address}/v1/${vault_mount.pki.path}/crl"]
+
+  depends_on = [vault_pki_secret_backend_root_cert.root_cert]
+}
+
+# Enable an intermediate PKI secrets engine.
+resource "vault_mount" "pki_int" {
+  path                      = "pki_int"
+  type                      = "pki"
+  default_lease_ttl_seconds = 3600
+  max_lease_ttl_seconds     = 311040000
+  description               = "PKI intermediate secrets engine"
+
+  depends_on = [vault_pki_secret_backend_config_urls.pki_urls]
+}
+
+# Create a Certificate Signing Request (CSR) for the intermediate CA.
+resource "vault_pki_secret_backend_intermediate_cert_request" "pki_int" {
+  backend = vault_mount.pki_int.path
+  type    = "internal"
+
+  common_name = "svc Intermediate Authority" 
+
+  depends_on = [vault_mount.pki_int]
+}
+
+# Store the CSR locally for signing.
+resource "local_file" "pki_intermediate_csr" {
+  content  = vault_pki_secret_backend_intermediate_cert_request.pki_int.csr
+  filename = "${path.module}/pki_intermediate.csr"
+
+  depends_on = [vault_pki_secret_backend_intermediate_cert_request.pki_int]
+}
+
+# Sign the intermediate CSR with the root CA to generate the intermediate certificate.
+resource "vault_pki_secret_backend_root_sign_intermediate" "pki_int_signed" {
+  backend = vault_mount.pki.path  # Root CA backend.
+  csr     = vault_pki_secret_backend_intermediate_cert_request.pki_int.csr
+
+  common_name = "svc Intermediate Authority"
+  ttl         = "43800h"  # TTL for the intermediate cert (~5 years).
+  format      = "pem_bundle"
+
+  depends_on = [
+    vault_pki_secret_backend_intermediate_cert_request.pki_int,
+    vault_mount.pki
+  ]
+}
+
+# Store the signed intermediate certificate locally.
+resource "local_file" "intermediate_cert_pem" {
+  content  = vault_pki_secret_backend_root_sign_intermediate.pki_int_signed.certificate
+  filename = "${path.module}/intermediate.cert.pem"
+
+  depends_on = [vault_pki_secret_backend_root_sign_intermediate.pki_int_signed]
+}
+
+# Set the signed intermediate certificate in the 'pki_int' backend.
+resource "vault_pki_secret_backend_intermediate_set_signed" "pki_int" {
+  backend     = vault_mount.pki_int.path
+  certificate = vault_pki_secret_backend_root_sign_intermediate.pki_int_signed.certificate
+
+  depends_on = [vault_pki_secret_backend_root_sign_intermediate.pki_int_signed]
+}
+
+# Configure the PKI URLs for 'pki_int'.
+resource "vault_pki_secret_backend_config_urls" "pki_int_urls" {
+  backend = vault_mount.pki_int.path
+
+  issuing_certificates    = ["${var.vault_address}/v1/${vault_mount.pki_int.path}/ca"]
+  crl_distribution_points = ["${var.vault_address}/v1/${vault_mount.pki_int.path}/crl"]
+
+  depends_on = [vault_pki_secret_backend_intermediate_set_signed.pki_int]
+}
+
+# Enable AppRole authentication in Vault.
```

Comment:
Is there no way for the cert manager to login via ServiceAccount? That would be much nicer IMO compared to using AppRoles.

Reply:
it is eventually possible, but I am not sure if the one click setup would work. From my understanding, we would need to create the k3d cluster, get the kubernetes_host, the kubernetes_ca_cert and the token_reviewer_jwt. Do you think that it is worth it?

Reply:
mhhh CA is an issue I see. I have a crazy idea:

1. Bootstrap Vault via TF with configuring it.
2. Bootstrap K3d via TF using https://registry.terraform.io/providers/pvotal-tech/k3d/latest/docs/resources/cluster
3. Read out Kubernetes CA from data source using `yamldecode` on [`kubeconfig_raw`](https://registry.terraform.io/providers/pvotal-tech/k3d/latest/docs/data-sources/cluster#kubeconfig_raw)
4. Configure Vault with TF. Here we use the JWT authentication engine instead, since the JWKS endpoint is fixed as long as we know the Kubernetes API, which we can fix in the bootstrap of k3d. The only missing part is then the CA, which we have from step 3.
5. Bootstrap flux on the cluster using https://registry.terraform.io/providers/fluxcd/flux/latest/docs

Then we have the entire setup end to end using a single terraform apply :)